### PR TITLE
OAK-11184 : introduce prevNoProp cache to avoid expensive prev docs scan

### PR DESCRIPTION
**Pull request by stefan-egli**

### CODE REVIEW

**Comment by stefan-egli** [comment]
draft now contains both required parts of a fix:
* first part is simple : check for uncommitted state. if the local map contains only uncommitted revisions, then we don't need to scan previous documents. this state however happens only for a very small amount of time (between writing the changes and setting the revision committed)
* second part now done based on cache approach : introduce a tiny (1%) cache that remembers whether a previous document (and its children) had no revision whatsoever of a particular property. This can then be used to avoid subsequent scans. But the first scan is necessary in this approach. We could further restrict the cache to only be used on root and say first level children - to further reduce impact and to ensure it remains tiny indeed.

FTR: another approach would be to store an explicit null revision whenever adding a new property to a document (or restrict that to root and eg first level children). The null revision would have to be chosen in a way to ensure it is always visible. I.e. it probably would have to be something like "r1-0-i". The impact of this second approach is slightly wider, as it does change what's written to the DocumentStore - hence the first suggestion to discuss the cache approach.

**Comment by stefan-egli** [comment]
The suggestion is to go with the `prevNoProp` cache and hide it behind a feature toggle (disabled by default).

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java [hunk 1]
```diff
@@ -684,6 +691,13 @@ public int getMemory() {
 
         diffCache = builder.getDiffCache(this.clusterId);
 
+        final Feature prevNoPropFeature = builder.getPrevNoPropCacheFeature();
+        if (prevNoPropFeature == null || !prevNoPropFeature.isEnabled()) {
+            prevNoPropCache = null;
+        } else {
+            prevNoPropCache = builder.buildPrevNoPropCache(this);
+        }
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: I would move this to [Utils](https://github.com/apache/jackrabbit-oak/blob/trunk/oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/Utils.java) and add unit cases for the same.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 2]
```diff
@@ -608,13 +621,17 @@ public long getPrevDocumentCacheSize() {
         return memoryCacheSize * prevDocCachePercentage / 100;
     }
 
+    public long getPrevNoPropCacheSize() {
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:627]: Shouldn't this check feature toggle first ?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 3]
```diff
@@ -608,13 +621,17 @@ public long getPrevDocumentCacheSize() {
...
+    }
+
     public long getChildrenCacheSize() {
         return memoryCacheSize * childrenCachePercentage / 100;
     }
 
     public long getDocumentCacheSize() {
         return memoryCacheSize - getNodeCacheSize() - getPrevDocumentCacheSize() - getChildrenCacheSize()
-                - getDiffCacheSize();
+                - getDiffCacheSize() - getPrevNoPropCacheSize();
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:641]: same as above, value should depend upon prevNoPropCache feature toggle.

**Review by rishabhdaim — CHANGES_REQUESTED** [review]

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java [hunk 4]
```diff
@@ -671,6 +680,10 @@ protected void deactivate() {
             docStoreEmbeddedVerificationFeature.close();
         }
 
+        if (prevNoPropCacheFeature != null) {
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:684]: Though this is unrelated to this PR, could you please add a utility method that accepts an array of `Feature` and closes them all?

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java (hunk 1 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: good point, will do

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 2 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:627]: good point - if feature toggle is off it should actually use `0`

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:641]: indeed

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java (hunk 4 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:684]: I'd suggest to move this to a separate ticket as it is unrelated. +1 for changing it though (but it could happen with lower prio). Wdyt?

Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:684]: jira OAK-11248 created

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreService.java:684]: thx!

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java (hunk 1 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: moved to the builder actually, as a few things need to be checked, done in https://github.com/apache/jackrabbit-oak/pull/1779/commits/faa44dffb661a8776dcef7d335e239096e17fce4

(will add some more unit tests)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 2 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:627]: addressed in https://github.com/apache/jackrabbit-oak/pull/1779/commits/faa44dffb661a8776dcef7d335e239096e17fce4

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 3 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java:641]: addressed in https://github.com/apache/jackrabbit-oak/pull/1779/commits/faa44dffb661a8776dcef7d335e239096e17fce4 - also introduced an actual config (with default 1). toggle overwrites config of course.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java [hunk 5]
```diff
@@ -915,7 +922,25 @@ public NodeDocumentCache buildNodeDocumentCache(DocumentStore docStore, NodeDocu
...
+        final Feature prevNoPropFeature = getPrevNoPropCacheFeature();
+        return prevNoPropFeature != null && prevNoPropFeature.isEnabled();
+    }
+
+    /**
+     * Builds the prevNoProp cache, if the corresponding feature toggle is enabled.
+     * Returns null otherwise
+     * @return null if prevNoProp feature is disabled and size non-null, a newly built cache otherwise
+     */
+    public Cache<StringValue, StringValue> buildPrevNoPropCacheOrNull() {
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: IMO `buildPrevNoPropCache` should be sufficient with javadocs explaining that the return value could be `null`.

Also, we should annotates this method with `@Nullable`.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java (hunk 1 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStore.java]: more tests added in https://github.com/apache/jackrabbit-oak/pull/1779/commits/fd3f5f68b00f88d0a5a4c6e0c92affd95d3b398b
(also moved `createFeature` to `UtilsTest` for code reuse)

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java (hunk 5 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreBuilder.java]: true, done in https://github.com/apache/jackrabbit-oak/pull/1779/commits/d1e238578ae11092a46d078b34e0d0443364c381

**Comment by stefan-egli** [comment]
@rishabhdaim  fyi : done some more minor changes, mostly beautification. code should otherwise be ready for a potential other round of review.

**Comment by stefan-egli** [comment]
https://github.com/apache/jackrabbit-oak/pull/1779/commits/4965c17ee47e99eb9ca364ffd7911b940722de6a is an additional precaution : only do any kind of filtering (for uncommitted or not visible) if prevNoProp feature toggle is enabled (checked indirectly by requiring the cache to be not null)

**Review by reschke — COMMENTED** [review]
I can't really tell, but if we can disable it...

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java [hunk 6]
```diff
@@ -106,6 +106,11 @@
             description = "Percentage of cache to be allocated towards Diff cache")
     int diffCachePercentage() default DEFAULT_DIFF_CACHE_PERCENTAGE;
 
+    @AttributeDefinition(
+            name = "PrevNoProp Cache",
+            description = "Percentage of cache to be allocated towards PrepNoProp cache")
```
Comment (rishabhdaim) [inline oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]:
```suggestion
            description = "Percentage of cache to be allocated towards PrevNoProp cache")
```

Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: Also, a one-liner description about this cache would be a good addition.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java [hunk 7]
```diff
@@ -188,6 +196,255 @@ public void blockingBlob() throws Exception {
...
+        DocumentNodeStore ns1 = builderProvider.newBuilder().setClusterId(1).setAsyncDelay(0).clock(clock)
+                .setPrevNoPropCacheFeature(createFeature(prevNoPropCacheEnabled))
+                .setDocumentStore(store1).build();
+
+        NodeBuilder b1 = ns1.getRoot().builder();
+        b1.setProperty("prop", -1);
+        ns1.merge(b1, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+
+        // create MANY previous docs
+        System.out.println(new Date() + " - creating MANY previous docs...");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: please remove this sout.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java [hunk 8]
```diff
@@ -188,6 +196,255 @@ public void blockingBlob() throws Exception {
...
+        for (int j = 0; j < 100; j++) {
+            for (int i = 0; i < NodeDocument.NUM_REVS_THRESHOLD; i++) {
+                b1 = ns1.getRoot().builder();
+                b1.setProperty("prop", i);
+                ns1.merge(b1, EmptyHook.INSTANCE, CommitInfo.EMPTY);
+            }
+            ns1.runBackgroundOperations();
+        }
+        System.out.println(new Date() + " - done creating MANY previous docs.");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: same as above.

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java [hunk 9]
```diff
@@ -188,6 +196,255 @@ public void blockingBlob() throws Exception {
...
+        store1.pauseWith(pauseCallback).on(NODES).on("0:/").after(2).eternally();
+        mergeThread.start();
+        boolean breakpointReached = breakpointReachedSemaphore.tryAcquire(5, TimeUnit.SECONDS);
+        assertTrue(breakpointReached);
+
+        try {
+            // start B
+            // for MongoDocumentStore (and perhaps others), we can't use the same
+            // instance by the second DocumentNodeStore, as that would mean they
+            // share the cache. And it is also the cache that is part of teh testing.
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]:
```suggestion
            // share the cache. And it is also the cache that is part of the testing.
```

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java [hunk 10]
```diff
@@ -188,6 +196,255 @@ public void blockingBlob() throws Exception {
...
+                // this worked irrespective of the fix:
+                assertNoPreviousDocsRead(c, 95);
+            }
+        } finally {
+            // in case anyone is still waiting (eg in a test failure case) :
+            continueSemaphore.release();
+        }
+
+        // a bit simplistic, but that's one way to reproduce the bug
+        System.out.println(new Date() + " - success.");
```
Comment (rishabhdaim) [inline oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: please remove this as well.

**Review by rishabhdaim — CHANGES_REQUESTED** [review]
added some minor review comments.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 6 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: thx, applied - will do one-liner next

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java (hunk 7 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: don't like system outs? :)

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java (hunk 10 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: ok ok ..

Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: removed in https://github.com/apache/jackrabbit-oak/pull/1779/commits/6421cd87c101ffef3795c5f2e928ce48a4ae9e5b

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java (hunk 8 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: removed in https://github.com/apache/jackrabbit-oak/pull/1779/commits/6421cd87c101ffef3795c5f2e928ce48a4ae9e5b

File: oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java (hunk 7 above)
Comment (stefan-egli) [reply oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/DocumentNodeStoreIT.java]: removed in https://github.com/apache/jackrabbit-oak/pull/1779/commits/6421cd87c101ffef3795c5f2e928ce48a4ae9e5b

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 6 above)
Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: where should the one-liner go? we usually aren't doing a lot of description of the caches in the config value themselves neither..

Comment (stefan-egli) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: added something here : https://github.com/apache/jackrabbit-oak/pull/1779/commits/9b055cb59294f854e84a8ac36ba828c3fa3dfe12

**Review by Joscorbe — APPROVED** [review]
Looks good to me.

File: oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java (hunk 6 above)
Comment (rishabhdaim) [reply oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/Configuration.java]: I agree, we don't have many descriptions on the other caches as well. let's skip this one too.

**Review by rishabhdaim — APPROVED** [review]

**Comment by stefan-egli** [comment]
two after-vote changes : https://github.com/apache/jackrabbit-oak/pull/1779/commits/10a5fda46ba6cbf1f222b4ee9e8cca41c5bd7768 with a logger rename and the minor https://github.com/apache/jackrabbit-oak/pull/1779/commits/12c8399460679c2fa16a454496d9b77aa5f1d3de with only comment update.

**Comment by stefan-egli** [comment]
.. and another minor javadoc one : https://github.com/apache/jackrabbit-oak/pull/1779/commits/65a0ff8f69146f7972d7ba9e5b4c2a385f4ac7a0

slowly getting there ...

**Comment by stefan-egli** [comment]
and not entirely unexpected : https://github.com/apache/jackrabbit-oak/pull/1779/commits/7f577131cb7c3abcbe0ba4a6132a57adbf07c90e with a test setup fix

**Comment by stefan-egli** [comment]
Some flakyness in DocumentNodeStoreIT with RDB - will disable one of the new tests to get this PR merged
```
[ERROR] Tests run: 12, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 66.495 s <<< FAILURE! - in org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT
[ERROR] unmergedCommitOnRoot_withoutPrevNoPropCache[RDBFixture: RDB-H2(file)](org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT)  Time elapsed: 0.742 s  <<< ERROR!
java.lang.Exception: Unexpected exception, expected<java.lang.AssertionError> but was<org.apache.jackrabbit.oak.plugins.document.DocumentStoreException>
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:30)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
Caused by: org.apache.jackrabbit.oak.plugins.document.DocumentStoreException: Configured cluster node id 1 already in use: leaseEnd 1730834632910 > 1730834454039 - 178871ms in the future
	at org.apache.jackrabbit.oak.plugins.document.ClusterNodeInfo.createInstance(ClusterNodeInfo.java:683)
	at org.apache.jackrabbit.oak.plugins.document.ClusterNodeInfo.getInstance(ClusterNodeInfo.java:511)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStore.<init>(DocumentNodeStore.java:620)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBuilder.build(DocumentNodeStoreBuilder.java:196)
	at org.apache.jackrabbit.oak.plugins.document.DocumentMKBuilderProvider$DisposingDocumentMKBuilder.build(DocumentMKBuilderProvider.java:82)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT.doUnmergedCommitOnRoot(DocumentNodeStoreIT.java:239)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT.unmergedCommitOnRoot_withoutPrevNoPropCache(DocumentNodeStoreIT.java:226)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.ExpectException.evaluate(ExpectException.java:19)
	... 31 more
 
[ERROR] modifiedResetWithDiff[RDBFixture: RDB-H2(file)](org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT)  Time elapsed: 0.335 s  <<< ERROR!
org.apache.jackrabbit.oak.plugins.document.DocumentStoreException: Configured cluster node id 1 already in use: leaseEnd 1730834632910 > 1730834454042 - 178868ms in the future
	at org.apache.jackrabbit.oak.plugins.document.ClusterNodeInfo.createInstance(ClusterNodeInfo.java:683)
	at org.apache.jackrabbit.oak.plugins.document.ClusterNodeInfo.getInstance(ClusterNodeInfo.java:511)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStore.<init>(DocumentNodeStore.java:620)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreBuilder.build(DocumentNodeStoreBuilder.java:196)
	at org.apache.jackrabbit.oak.plugins.document.DocumentNodeStoreIT.modifiedResetWithDiff(DocumentNodeStoreIT.java:95)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:384)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:345)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:126)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:418)
 ```

**Comment by stefan-egli** [comment]
disabled https://github.com/apache/jackrabbit-oak/pull/1779/commits/ad54439f967a33342ecc512f531185f30a2c4b23

otherwise tests succeeded, hence merging
